File: java/src/main/HouseRobberII.java
import static util.ArrayUtils.array;

/**
 * 213. 打家劫舍 II：https://leetcode-cn.com/problems/house-robber-ii/
 */
public class HouseRobberII {

    public static int rob(int[] nums) {
        int n = nums.length;
        if (n <= 2) {
            return n == 1 ? nums[0] : Math.max(nums[0], nums[1]);
        }
        // 0 ~ n - 1
        int first = nums[0], second = Math.max(nums[0], nums[1]);
        for (int i = 2; i < n - 1; i++) {
            int t = second;
            second = Math.max(first + nums[i], second);
            first = t;
        }
        int ans = second;

        // 1 ~ n
        first = nums[1];
        second = Math.max(nums[1], nums[2]);
        for (int i = 3; i < n; i++) {
            int t = second;
            second = Math.max(first + nums[i], second);
            first = t;
        }
        ans = Math.max(ans, second);
        return ans;
    }

    public static void main(String[] args) {
        System.out.println(rob(array("[2,3,2]"))); // 3
        System.out.println(rob(array("[1,2,3,1]"))); // 4
        System.out.println(rob(array("[1,2,3]"))); // 3
    }

}

File: c/house-robber-ii.c
/**
 * 213. 打家劫舍 II：https://leetcode.cn/problems/house-robber-ii
 */

#include <stdio.h>

#define MAX(a, b) ((a) > (b) ? (a) : (b))

int rob1(int* nums, int start, int end) {
    int f0 = 0, f1 = 0;
    for (int i = start; i <= end; i++) {
        int f = MAX(nums[i] + f0, f1);
        f0 = f1;
        f1 = f;
    }
    return f1;
}

int rob(int* nums, int n) {
    return MAX(nums[0] + rob1(nums, 2, n - 2), rob1(nums, 1, n - 1));
}

int main(int argc, char *argv[]) {
    int nums[] = {2,3,2};
    printf("%d\n", rob(nums, 3));
    int nums2[] = {1,2,3,1};
    printf("%d\n", rob(nums2, 4));
    int nums3[] = {1,2,3};
    printf("%d\n", rob(nums3, 3));
}
